### PR TITLE
Fall back to readBinary/readAsync in dlopen when file not found in FS

### PR DESCRIPTION
**Pull request by sbc100**
If the file isn't found in the filesystem we can always
fall back to loading from a URL.

Also, remove the redundant/duplicate function `fetchBinary` which
is just promisified version of what `readAsync` already does.

Also, since I've added `readAsync` for nodejs we remove the
extra codepath that existed only when `readBinary` existed
and just always used `readAsync`.

Split out from #14503

### CODE REVIEW

File: src/library_dylink.js [hunk 1]
```diff
@@ -637,9 +625,13 @@ var LibraryDylink = {
...
 
       if (flags.loadAsync) {
-        return fetchBinary(libFile);
+        return new Promise(function(resolve, reject) {
+          readAsync(libFile, function(data) { resolve(new Uint8Array(data)); }, reject);
+        });
       }
+
       // load the binary synchronously
+      if (!readBinary) throw new Error('file not found');
```
Comment (kripken) [inline src/library_dylink.js:634]:
```suggestion
      if (!readBinary) {
        throw new Error(`file ${libfile} not available, and no way to read it synchronously`);
      }
```

File: src/library_dylink.js [hunk 2]
```diff
@@ -694,32 +686,24 @@ var LibraryDylink = {
...
+#if DYLINK_DEBUG
+    err('preloadDylibs done!');
+#endif
     });
-    reportUndefinedSymbols();
   },
 
   // void* dlopen(const char* filename, int flags);
   dlopen__deps: ['$DLFCN', '$FS', '$ENV'],
-  dlopen__proxy: 'sync',
```
Comment (kripken) [inline src/library_dylink.js:722]: why are these changed here?

File: tests/test_browser.py [hunk 3]
```diff
@@ -3538,6 +3538,15 @@ def test_dynamic_link(self):
     # same wasm side module works
     self.btest_exit(self.in_dir('main.c'), args=['-s', 'MAIN_MODULE=2', '-O2', '-s', 'EXPORT_ALL', 'side.wasm'])
 
+  def test_dlopen_blocking(self):
+    create_file('side.c', 'int foo = 42;\n')
+    self.run_process([EMCC, 'side.c', '-o', 'libside.so', '-s', 'SIDE_MODULE', '-s', 'USE_PTHREADS', '-Wno-experimental'])
```
Comment (kripken) [inline tests/test_browser.py:3543]: why "no-experimental"?

File: tests/test_browser.py [hunk 4]
```diff
@@ -3538,6 +3538,15 @@ def test_dynamic_link(self):
...
+    # Attempt to use dlopen without preloading the side module should fail on the main thread
+    # since the syncronous `readBinary` function does not exist.
+    self.btest_exit(test_file('other/test_dlopen_blocking.c'), assert_returncode=1, args=['-s', 'MAIN_MODULE=2'])
+    # But with PROZY_TO_PTHEAD it does work since we can block.
```
Comment (kripken) [inline tests/test_browser.py:3547]:
```suggestion
    # But with PROXY_TO_PTHEAD it does work since we can block.
```

File: tests/test_browser.py [hunk 5]
```diff
@@ -3538,6 +3538,15 @@ def test_dynamic_link(self):
     # same wasm side module works
     self.btest_exit(self.in_dir('main.c'), args=['-s', 'MAIN_MODULE=2', '-O2', '-s', 'EXPORT_ALL', 'side.wasm'])
 
+  def test_dlopen_blocking(self):
+    create_file('side.c', 'int foo = 42;\n')
+    self.run_process([EMCC, 'side.c', '-o', 'libside.so', '-s', 'SIDE_MODULE', '-s', 'USE_PTHREADS', '-Wno-experimental'])
+    # Attempt to use dlopen without preloading the side module should fail on the main thread
+    # since the syncronous `readBinary` function does not exist.
+    self.btest_exit(test_file('other/test_dlopen_blocking.c'), assert_returncode=1, args=['-s', 'MAIN_MODULE=2'])
```
Comment (kripken) [inline tests/test_browser.py:3546]: how about adding a test with preloading?

File: tests/test_browser.py (hunk 3 above)
Comment (sbc100) [reply tests/test_browser.py:3543]: Because dynamic linking + threading generates a warning otherwise.

File: src/library_dylink.js (hunk 2 above)
Comment (sbc100) [reply src/library_dylink.js:722]: Otherwise the dlopen will get proxied back to then the main thread.. where it can't block.

The list of loaded modules is worker specific so we don't want to be proxying dynamic loading calls back to the main thread.

File: src/library_dylink.js (hunk 1 above)
Comment (sbc100) [reply src/library_dylink.js:634]: This is tricky one.  Previously, the user would just have see the "file not found" error from the filsystem and we would never have got here.    I will re-phrase.

File: tests/test_browser.py (hunk 5 above)
Comment (sbc100) [reply tests/test_browser.py:3546]: All of the existing dlopen tests use pre-loading so I think we already have plenty of coverage of that.

File: src/library_dylink.js [hunk 6]
```diff
@@ -694,32 +688,24 @@ var LibraryDylink = {
...
+    // Load binaries asynchronously
+    addRunDependency('preloadDylibs');
+    dynamicLibraries.reduce(function(chain, lib) {
+      return chain.then(function() {
+        return loadDynamicLibrary(lib, {loadAsync: true, global: true, nodelete: true, allowUndefined: true});
       });
-      return;
-    }
-
-    dynamicLibraries.forEach(function(lib) {
```
Comment (dschuff) [inline src/library_dylink.js:713]: I don't see this logic anywhere in the new version, is that intentional?

Comment (sbc100) [reply src/library_dylink.js:713]: Yes, now that readAsync, is ubiquitous we can just have single code path there that works everywhere.  No need for the extra code path above.

File: src/library_dylink.js (hunk 2 above)
Comment (kripken) [reply src/library_dylink.js:722]: Makes sense, except don't we (eventually?) actually want a central authority to dlopen things, and communicate that to the other threads?

**Review by kripken — APPROVED** [review]

File: src/library_dylink.js (hunk 2 above)
Comment (sbc100) [reply src/library_dylink.js:722]: Eventually we will need some kind of co-ordination to keep the threads in sync.   However, I imagine that the solution we use will still involve each thread having its own state (since each thread has its own wasmTable which could be out of sync with the other threeds).   Exactly how the threads are kept in sync of catch up with each other is still an open question.   

Also, it it seems like the thread that calls `dlopen` will probably want to load the dll in its own context right away.
